### PR TITLE
[19685] network rate limit: allow limiting by client IP

### PR DESCRIPTION
**Pull request by llu94**
Commit Message: Dynamic injection of downstream ip into remote address descriptor entry in the L4 Network Rate Limit Filter. 
Additional Description: See this issue -> https://github.com/envoyproxy/envoy/issues/19685
Risk Level: Low
Testing: Unit Testing
Docs Changes: Required before merge
Release Notes: N/A
Platform Specific Features: N/A
[Optional Fixes #Issue] - #19685

### CODE REVIEW

**Comment by llu94** [comment]
Currently a WIP. Need to have a convo with the code owner and requester over features and implementation details. 

Will write up documentation once finished and before merge.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (daixiang0) [inline source/extensions/filters/network/ratelimit/ratelimit.h:101]: I think we can do it step by step, if so we can reduce changes and make review easy.

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:101]: @daixiang0 so you're saying just get a happy path/simple solution in first and iterate on it in later PRs?

Comment (daixiang0) [reply source/extensions/filters/network/ratelimit/ratelimit.h:101]: /cc @envoyproxy/envoy-maintainers WDYT?

Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.h:101]: It's fine to start with a simple approach, however once we have certain implicit behavior we can't change it, so we need to be careful to make sure whatever we do is future proof.

One thing I would like to understand is whether we should be using the substituion formatter here with a limited set of supported substitions. I think that would be better.

So basically the key is generic and set by the user, the value is optionally a substituion format string. Can we look at that and try to implement it that way?

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:100]: I think this implementation will change but no C string handling, use C++/absl/etc.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this. Some high level comments to get started.

/wait

**Comment by llu94** [comment]
@mattklein123 in the interest of saving me some time, is there an example of the substitution formatter approach that I can look at?

**Comment by rapatchi** [comment]
@llu94  thanks for taking a look. If possible could you also add in the description how this will finally translate in envoy's configuration file ?.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (yanavlasov) [reply source/extensions/filters/network/ratelimit/ratelimit.h:101]: Here is an example of existing formatters: https://github.com/envoyproxy/envoy/blob/main/source/common/formatter/substitution_formatter.h

This will allow you to specify the descriptor value in a form %REMOTE_IP%. In this way it is a bit more flexible than using predefined key names.

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:100]: @mattklein123 @yanavlasov @rapatchi I have added some proto definitions to the API to accommodate the substitution formatters in the form of a new set of descriptors explicitly for the network rate limiter. Please let me know if this is workable or what changes need to be made. I'd like to have the API definition agreed on before I commit the time to implement this and write tests.

File: api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto:42]: Do we really need a new message here? Can't we just put a substitution format string directly inside the existing descriptor? I admit this could theoretically break someone, but the chance of that in practice is basically zero, and we can runtime feature flag it to be safe?

**Comment by llu94** [comment]
@rapatchi @mattklein123 made some changes and added some preliminary documentation. Apologies for the slow pace of updates, real life struck.

File: api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto
Comment (llu94) [reply api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto]: Change is in. Should this be in the snapshot or no?

File: api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto:35]: This will need a release note and more complete documentation in the filter

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:106]: The way this should work is the entire value should be fed into a substitution formatter, much like we do for access logging. Any value that the formatter supports should work, as long as stream info is populated. Essentially, almost any value that tcp_proxy supports should work here. This will make it much more general purpose.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thank's a ton for working on this. This is on the right track. Flushing out some more comments.

/wait

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h]: I've changed the implementation to a substitution formatter based approach. Can you take a look and see if this is what you envisioned? And what sort of documentation you'd like to see?

File: api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto:35]: Ping on this. Please add docs and a release note.

File: api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto:57]: I don't think this is needed. The idea here is to just support this by default. If you want to guard it you can use a RuntimeFeatureFlag and document that.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:4]: ping same below

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:106]: You shouldn't need to do any of this. The idea is to pre-create all of the formatters when you read the config, and then just apply them directly during proxying.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks left a few more comments.

/wait

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (ggreenway) [reply source/extensions/filters/network/ratelimit/ratelimit.h:106]: Just to clarify: what we want to do is allow the user to add substitution strings to the existing `RateLimitDescriptor.Entry.value` in their filter config, and the only change is that we run `substitutionFormattedString` on those now?

Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.h:106]: Yes, sorry for not being more clear.

The idea is that we should parse the entry value into a substitution formatter on configuration load. The common case will be a string -> string mapping which will have negligible perf impact. Otherwise, we can use the formatter when populating the real entry that gets sent to the server given the current stream info in the filter chain. Let me know if you need more code pointers.

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:106]: @mattklein123 the stream_info used to scrape the information isn't initialized in the config. I do understand that we just want to generally apply the substitution formatter to these descriptors now though.

Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.h:106]: Yes, stream info is supplied at request time, not at config load time, into the pre created formatter.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks looks generally correct. Can you please clean this up and we can do final review? This needs docs, release notes, code cleanup, etc. Thank you.

/wait

**Comment by llu94** [comment]
> Thanks looks generally correct. Can you please clean this up and we can do final review? This needs docs, release notes, code cleanup, etc. Thank you.
> 
> /wait

I've cleaned up the code and have written docs and release notes. For future reference, where is the auto-formatter for the release notes/docs?

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (zirain) [inline source/extensions/filters/network/ratelimit/ratelimit.h:115]: can this also implement on http ratelimit?

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:115]: Did you mean can we implement this in the HTTP rate limiter?

Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.h:115]: Yes, optimally we would do this change in both places, but we could do this in a follow up.

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:115]: I'd prefer to do that in a followup. I can do that after this one.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:113]: The way this is being done is fairly inefficient. You are rebuilding the formatter every time you want to apply a format.  The way to do this is to build the formatter at configuration load time, and then you can apply it for each connection. So essentially, the formatter build code should live in the config object, and then you can consume it from the filter.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:111]: For perf reasons I would create some static headers and trailers for use to avoid construction every time you need to do this. These can live in the config object also per my below comment so that when you do the formatting through the config you can just use the static values.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:92]: nit: after refactor per below, move these function bodies into cc files, there is no need to have them in the header files.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:108]: s/body/descriptor_value or something

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks flushing out some more comments.

/wait

**Comment by mattklein123** [comment]
The ASAN failure is real. Please check CI.

/wait

**Comment by daixiang0** [comment]
Sorry for mistake :(

**Comment by llu94** [comment]
I'll see what's wrong. It's working locally so I thought it was okay, my bad.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:55]: Again, the idea is to pre-create these at config load time. Then you store a vector of these, and use them during apply.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:40]:
```suggestion
  std::vector<RateLimit::Descriptor> dynamic_descriptors = std::vector<RateLimit::Descriptor>();
```

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:55]: I'm fine with all of the other changes but I think that the performance hit for creating the Formatter and cleaning it out each substitution is not noticeable relative to having to keep a vector filled with formatter objects for especially complex and verbose rate limiting schemes. That being said I'm not as familiar with this code base as you are so I will defer. 

Making the other changes rn.

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:54]: @mattklein123 the FormatterImpl class doesn't have the copy functionality/relevant interfaces needed to be used with vectors it looks like. Should we change the FormatterImpl?

Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:54]: No, you can store them in a pointer if needed.

/wait

**Comment by llu94** [comment]
Remote provider keeps canceling. Attempting to retest.

**Comment by llu94** [comment]
Somehow one of the commits I made didn't do a correct GPG signoff? Fixing.

**Comment by rapatchi** [comment]
@llu94  Could you let me know what's the state of this PR ?. Wanted to consume it on our side and test it accordingly.

**Comment by llu94** [comment]
> @llu94 Could you let me know what's the state of this PR ?. Wanted to consume it on our side and test it accordingly.

It's finished. The remote server is saying it's out of resources though so I can't run the full remote test suite.

**Comment by mattklein123** [comment]
@llu94 please do not force push. It makes reviews more difficult. Just add commits and merge main. Thanks!

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:65]: nit: I think you can use unique_ptr here.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:115]: ? Please clean up the PR for things like this.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:57]: ?

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:44]: I don't quite understand what this function is doing. I think all you need to do is run through the formatters and apply all of them.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:65]: Other similar use cases in this project are using shared_ptr instead of unique_ptr and I"m not the best C++ developer as you can tell. I will keep as is.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:44]: It's applying each unique Formatter stored in the vector created during config in matched order as the descriptor values are stored once the stream_info is available.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.h:65]: I'm pretty sure you can use unique_ptr here, so let's give that a try.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:44]: See my above comment. You should only need to store the formatters and then run through them.

/wait

Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.cc:44]: The limit function requires a descriptor as input. That would have to change for this change to go through. Do we want to do that.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (llu94) [reply source/extensions/filters/network/ratelimit/ratelimit.h:64]: The client limit function requires a descriptor as input. That would have to change for this change to go through and affect everything using Filters::Common::RateLimit::ClientPtr. Do we want to do that.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:52]: delete

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:53]: nit: you can remove `Config::` here and below.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:51]: this line is not needed, just define and assign when you call format

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:41]: Per other comments, this function needs to return the final descriptors, not modify the config inline, which is not thread safe. Also, you don't need to pass original_descriptors, you already have it within the config.

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:44]:
```suggestion
  std::vector<RateLimit::Descriptor> dynamic_descriptors;
  dynamic_descriptors.reserve(original_descriptors_.size());
```

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:49]:
```suggestion
    for (const RateLimit::DescriptorEntry& descriptor_entry : descriptor.entries_) {
```

File: source/extensions/filters/network/ratelimit/ratelimit.cc
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.cc:48]:
```suggestion
    RateLimit::Descriptor new_descriptor;
    new_descriptor.entries_.reserve(descriptor.entries_.size());
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
/wait

**Comment by mattklein123** [comment]
Also revert the change to `[rst_check.py](https://github.com/envoyproxy/envoy/pull/19875/files#diff-6af1011ae2bfa4f15f2bb20647e44e7c714b68807ecc96ac0b484970a30766ce)`

**Comment by llu94** [comment]
> Also revert the change to `[rst_check.py](https://github.com/envoyproxy/envoy/pull/19875/files#diff-6af1011ae2bfa4f15f2bb20647e44e7c714b68807ecc96ac0b484970a30766ce)`

I think this is an issue downstream of some of the things I did to fix the sign off errors. What I have is a verbatim copy of what is in the main branch.

File: source/extensions/filters/network/ratelimit/ratelimit.h
Comment (mattklein123) [inline source/extensions/filters/network/ratelimit/ratelimit.h:103]: revert

**Review by mattklein123 — COMMENTED** [review]
Thanks LGTM modulo @yanavlasov comments.

/wait

**Comment by llu94** [comment]
Follow on issue for formatters without headers/http components is here: https://github.com/envoyproxy/envoy/issues/20818

**Comment by llu94** [comment]
Thanks for the patience and feedback on this everyone.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
